Add support for B3 parentspanid

ocelotl opened the pull request:
Fixes #236

ocelotl commented:
Hello

I think the CI failure is related to lint, specifically E1137. It kinda looks like a false positive, it has been reported [before](https://stackoverflow.com/questions/55986364/pylint-doesnt-seem-to-work-with-dicts-defined-in-dataclasses) for circumstances similar to this one:

```
opentelemetry-sdk/src/opentelemetry/sdk/context/propagation/b3_format.py:106:12: E1137: 'trace_state' does not support item assignment (unsupported-assignment-operation)
```
[Here](https://github.com/open-telemetry/opentelemetry-python/pull/286/commits/0cec30a936361078a2fbb018e8e9b18b892c767e#diff-9573b28f90915e5a64a091c1e74aa6a5R106) is the offending line.

ocelotl commented:
I signed it

Oberon00 commented:
This is arguably a true positive, since the trace state is supposed to be immutable. Of course it is implemented with a `dict` so it would work at runtime.

ocelotl commented:
Ok! I'll try that, @Oberon00 :+1:

ocelotl commented:
When I introduced this change:

```
diff --git a/opentelemetry-api/src/opentelemetry/trace/__init__.py b/opentelemetry-api/src/opentelemetry/trace/__init__.py
index 412a106..3c430b1 100644
--- a/opentelemetry-api/src/opentelemetry/trace/__init__.py
+++ b/opentelemetry-api/src/opentelemetry/trace/__init__.py
@@ -281,7 +281,9 @@ class TraceOptions(int):
 DEFAULT_TRACE_OPTIONS = TraceOptions.get_default()
 
 
-class TraceState(typing.Dict[str, str]):
+# dict is added here to the base classes of TraceState as a workaround for
+# pylint E1137 which is raised if dict is not present here.
+class TraceState(typing.Dict[str, str], dict):
     """A list of key-value pairs representing vendor-specific trace info.
 
     Keys and values are strings of up to 256 printable US-ASCII characters.
```

`tox -e lint` passed , but `tox -e py37-mypy` didn't:

```
py37-mypy installed: backcall==0.1.0,decorator==4.4.1,ipdb==0.12.2,ipython==7.9.0,ipython-genutils==0.2.0,jedi==0.15.1,mypy==0.740,mypy-extensions==0.4.3,parso==0.5.1,pexpect==4.7.0,pickleshare==0.7.5,prompt-toolkit==2.0.10,ptyprocess==0.6.0,Pygments==2.4.2,six==1.13.0,traitlets==4.3.3,typed-ast==1.4.0,typing-extensions==3.7.4.1,wcwidth==0.1.7
py37-mypy run-test-pre: PYTHONHASHSEED='1437297989'
py37-mypy run-test-pre: commands[0] | python -m pip install -U pip setuptools wheel
Requirement already up-to-date: pip in ./.tox/py37-mypy/lib/python3.7/site-packages (19.3.1)
Requirement already up-to-date: setuptools in ./.tox/py37-mypy/lib/python3.7/site-packages (41.6.0)
Requirement already up-to-date: wheel in ./.tox/py37-mypy/lib/python3.7/site-packages (0.33.6)
py37-mypy run-test: commands[0] | mypy --namespace-packages opentelemetry-api/src/opentelemetry/
opentelemetry-api/src/opentelemetry/trace/__init__.py:286: error: Duplicate base class "dict"
opentelemetry-api/src/opentelemetry/trace/__init__.py:286: error: Implicit generic "Any". Use "typing.Dict" and specify generic parameters
Found 2 errors in 1 file (checked 23 source files)
ERROR: InvocationError for command /home/ocelotl/repos/lightstep/opentelemetry-python/.tox/py37-mypy/bin/mypy --namespace-packages opentelemetry-api/src/opentelemetry/ (exited with code 2)
__________________________________________________________________________________ summary __________________________________________________________________________________
ERROR:   py37-mypy: commands failed
```
So, I updated this commit to simply ignore `E1137` in [that particular line](https://github.com/open-telemetry/opentelemetry-python/pull/286/commits/e727db1ad463b17d3d3460fe0d6250b63e11c561#diff-9573b28f90915e5a64a091c1e74aa6a5R106-R109). As reported by @Oberon00 before, this seems to be a [Pylint issue](https://github.com/PyCQA/pylint/issues/2420) and this approach just targets the very specific place where a workaround is needed.

ocelotl commented:
Ok, implemented the fix, please let me know if this is the right approach.

It fails `tox -e py3X-test-example-app` because of [this](https://github.com/open-telemetry/opentelemetry-python/blob/09f6c351b143034d27bcc36f849da0995ad24b2f/opentelemetry-api/src/opentelemetry/propagators/__init__.py#L67).

How should this be handled? That interface should also receive a `Span`? I tried making that line inject the span instead of its context but other test cases fail in that case. How should the interface be defined?

toumorokoshi commented on opentelemetry-sdk/src/opentelemetry/sdk/context/propagation/b3_format.py:
```diff
@@ -100,13 +101,22 @@ def extract(cls, get_from_carrier, carrier):
+        set_in_carrier(
+            carrier,
+            cls.PARENT_SPAN_ID_KEY,
+            format_span_id(span.parent.context.span_id),
```
the parent can be empty, so there should be a conditional here to check that first.

ocelotl replied:
Ok, added the check :+1:

toumorokoshi commented on opentelemetry-sdk/tests/context/propagation/test_b3_format.py:
```diff
@@ -35,22 +35,35 @@ def setUpClass(cls):
+            trace.generate_span_id()
+        )
 
     def test_extract_multi_header(self):
```
it feels to me that this test was sufficient by itself (since it exercises the base case of serialization / de-serialization), and have a separate test that verifies the parent_id propagation.

This test as is would also have been a good test for an empty parent.

ocelotl replied:
Sorry, I don't understand this comment, can you explain further, please?

toumorokoshi replied:
Yep definitely. What I was trying to say was: this test already should just work, and actually exercises a valuable case where the span has no parent.

So I was arguing to not modify this test case at all, and instead add another one that verifies that a parent_id field is set if a span has a parent.

ocelotl replied:
Sure, since you already approved these changes, @toumorokoshi, do you prefer if I open separate issues for your pending requests like this one or is if ok for you if I commit them directly in this PR again?

toumorokoshi commented on opentelemetry-sdk/src/opentelemetry/sdk/context/propagation/b3_format.py:
```diff
@@ -100,13 +101,22 @@ def extract(cls, get_from_carrier, carrier):
     @classmethod
-    def inject(cls, context, set_in_carrier, carrier):
-        sampled = (trace.TraceOptions.SAMPLED & context.trace_options) != 0
+    def inject(
```
this interface change will also need to be propagated up to the API layer as well. There are multiple propagators (tracecontext) that implement the previous API, and those will need to be changed as part of this.

ocelotl replied:
fixed too :+1:

toumorokoshi commented on opentelemetry-sdk/tests/context/propagation/test_b3_format.py:
```diff
@@ -35,22 +35,35 @@ def setUpClass(cls):
+        parent_span = trace.Span(
+            "parent", FORMAT.extract(get_as_list, carrier)
+        )
+        child_span = trace.Tracer().start_span("child", parent=parent_span)
```
I believe you could just instantiate a raw Span object yourself from the SDK, rather than use a whole trace to do so.

ocelotl replied:
Yes, fixed!

toumorokoshi commented on opentelemetry-sdk/tests/context/propagation/test_b3_format.py:
```diff
@@ -100,9 +148,12 @@ def test_enabled_sampling(self):
+            parent_span = trace.Span(
+                "parent", FORMAT.extract(get_as_list, carrier)
+            )
+            child_span = trace.Tracer().start_span("child", parent=parent_span)
```
there's a lot of boilerplate in these tests now. Could they be refactored into a few test helper methods?

ocelotl replied:
Yes, refactored :+1:

ocelotl commented on opentelemetry-sdk/src/opentelemetry/sdk/context/propagation/b3_format.py:
```diff
@@ -100,14 +101,24 @@ def extract(cls, get_from_carrier, carrier):
-        sampled = (trace.TraceOptions.SAMPLED & context.trace_options) != 0
+    def inject(
+        cls, span, set_in_carrier, carrier
+    ):  # pylint: disable=arguments-differ
```
Fixed!

ocelotl commented on ext/opentelemetry-ext-opentracing-shim/src/opentelemetry/ext/opentracing_shim/__init__.py:
```diff
@@ -667,12 +668,14 @@ def inject(self, span_context, format, carrier):
+
         propagator.inject(
-            span_context.unwrap(), type(carrier).__setitem__, carrier
+            Span("", span_context.unwrap()), type(carrier).__setitem__, carrier
```
I changed `Span` for `DefaultSpan`. I'm just curious, why is that you don't find this to be beautiful?

c24t replied:
Reviewing again, the reason I think this is a kludge is that we're only using the `DefaultSpan` to smuggle the span context to the propagator. We're taking advantage of the fact that `DefaultSpan`s aren't exported here, but everywhere else they're used interchangeably with regular `Span`s, with the expectation that the export behavior is left up to the user.

If the signature hadn't changed to require this to be a span, we wouldn't normally create a span here. That looks to me like evidence that we shouldn't change the API in this way.

ocelotl commented:
@c24t yes, I think this _is_ a violation of the spec since inject will receive now a `Span`. In a previous conversation, @toumorokoshi referred to an upcoming big change in the API (I think #325) that makes this change ok. Please @toumorokoshi, correct me if I am wrong.
